pkg/apprepo/apprepos_handler_test.go

```diff
@@ -188,19 +196,45 @@ func TestAppRepositoryCreate(t *testing.T) {
 					t.Errorf("mismatch (-want +got):\n%s", cmp.Diff(want, got))
 				}
 
-				// When appropriate, ensure the expected secret is stored
+				// When appropriate, ensure the expected secret is stored.
 				if appRepoRequest.AppRepository.AuthHeader != "" {
-					requestSecret := secretForRequest(appRepoRequest, responseAppRepo)
-					requestSecret.ObjectMeta.Namespace = tc.kubeappsNamespace
+					expectedSecret := secretForRequest(appRepoRequest, responseAppRepo)
+					responseSecret, err := cs.CoreV1().Secrets(appRepoRequest.AppRepository.Namespace).Get(expectedSecret.ObjectMeta.Name, metav1.GetOptions{})
 
-					responseSecret, err := cs.CoreV1().Secrets(tc.kubeappsNamespace).Get(requestSecret.ObjectMeta.Name, metav1.GetOptions{})
 					if err != nil {
-						t.Errorf("expected data %v not present: %+v", requestSecret, err)
+						t.Errorf("expected data %v not present: %+v", expectedSecret, err)
 					}
 
-					if got, want := responseSecret, requestSecret; !cmp.Equal(want, got) {
+					if got, want := responseSecret, expectedSecret; !cmp.Equal(want, got) {
 						t.Errorf("mismatch (-want +got):\n%s", cmp.Diff(want, got))
 					}
+
+					// Verify the copy of the repo secret in in kubeapps is
+					// also stored if this is a per-namespace app repository.
+					kubeappsSecretName := kubeappsSecretNameForRepo(requestAppRepo.ObjectMeta.Name, requestAppRepo.ObjectMeta.Namespace)
+					expectedSecret.ObjectMeta.Name = kubeappsSecretName
+					expectedSecret.ObjectMeta.Namespace = tc.kubeappsNamespace
+
+					if requestAppRepo.ObjectMeta.Namespace != tc.kubeappsNamespace {
+						responseSecret, err = handler.svcKubeClient.CoreV1().Secrets(tc.kubeappsNamespace).Get(kubeappsSecretName, metav1.GetOptions{})
+						if err != nil {
+							t.Errorf("expected data %v not present: %+v", expectedSecret, err)
+						}
+
+						if got, want := responseSecret, expectedSecret; !cmp.Equal(want, got) {
+							t.Errorf("mismatch (-want +got):\n%s", cmp.Diff(want, got))
+						}
+					} else {
+						_, err = handler.svcKubeClient.CoreV1().Secrets(tc.kubeappsNamespace).Get(kubeappsSecretName, metav1.GetOptions{})
+						if statusErr, ok := err.(*errors.StatusError); ok {
+							status := statusErr.ErrStatus
+							if got, want := status.Code, int32(404); got != want {
+								t.Errorf("got: %d, want: %d", got, want)
+							}
+						} else {
+							t.Errorf("Unable to convert err to StatusError: %+v", err)
+						}
+					}
```

Comment:
this test body is becoming huge, I wonder if it would be better to start splitting the different checks in different tests

Reply:
Yeah, I thought the same thing. I'll look at this in the following PRs.